main/guides/zoe/contract-upgrade.md

```diff
@@ -0,0 +1,186 @@
+The `publicFacet` and `creatorFacet` are defined by the contract.
+The `adminFacet` is defined by Zoe and includes methods to upgrade the contract.
+
+::: tip Upgrade Governance
+
+Governance of the right to upgrade is a complex topic that we cover only briefly here.
+
+- When [BLD staker governance](https://community.agoric.com/t/about-the-governance-category/15) makes a decision to start a contract using [swingset.CoreEval](../coreeval/),
+  to date, the `adminFacet` is stored in the bootstrap vat, allowing
+  the BLD stakers to upgrade such a contract in a later `swingset.CoreEval`.
+- The `adminFacet` reference can be discarded, so that noone can upgrade
+  the contract from within the JavaScript VM. (BLD staker governace
+  could, in theory, change the VM itself.)
+- The `adminFacet` can be managed using the [@agoric/governance](https://github.com/Agoric/agoric-sdk/tree/master/packages/governance#readme) framework; for example, using the `committee.js` contract.
+
+:::
+
+Upgrading a contract instance means re-starting the contract using a different [code bundle](./#bundling-a-contract). Suppose we start a contract as usual, using
+the bundle ID of a bundle we already sent to the chain:
+
+```js
+const bundleID = 'b1-1234abcd...';
+const installation = await E(zoe).installBundleID(bundleID);
+const { instance, ... facets } = await E(zoe).startInstance(installation, ...);
+
+// ... use facets.publicFacet, instance etc. as usual
+```
+
+If we have the `adminFacet` and the bundle ID of a new version,
+we can use the `upgradeContract` method to upgrade the contract instance:
+
+```js
+const v2BundleId = 'b1-feed1234...`; // hash of bundle with new feature
+const { incarnationNumber } = await E(facets.adminFacet).upgradeContract(v2BundleId);
+```
+
+The `incarnationNumber` is 1 after the 1st upgrade, 2 after the 2nd, and so on.
+
+::: details re-using the same bundle
+
+Note that a "null upgrade" that re-uses the original bundle is valid, and a legitimate approach to deleting accumulated heap state.
+
+See also `E(adminFacet).restartContract()`.
+
+:::
+
+## Upgradable Contracts
+
+There are a few requirements for the contract that differ from non-upgradable contracts:
+
+1. [Upgradable Declaration](#upgradable-declaration)
+2. [Durability](#durability)
+3. [Kinds](#kinds)
+4. [Crank](#crank)
+
+### Upgradable Declaration
+
+The new code bundle declares that it supports upgrade by exporting a `prepare` function in place of `start`.
+
+<<< @/snippets/zoe/src/02b-state-durable.js#export-prepare
+
+### Durability
+
+The 3rd argument, `baggage`, of the `prepare` function is a `MapStore`
+that provides a way to preserve state and behavior of objects
+between incarnations in a way that preserves identity of objects
+as seen from other vats:
+
+```js
+let rooms;
+if (!baggage.has('rooms')) {
+  // initial incarnation: create the object
+  rooms = makeScalarBigMapStore('rooms', { durable: true });
```

Comment:
I like that you use zones to make durable exos below. But then you should also use zones to make durable stores. Both for consistency, and because it's a much better API --- without "Big" or `{ durable: true }`

Reply:
That's in there, right? This section is showing the fully elaborated code that's easier to understand the first time you see it. But then it's replaced by more and more concise idioms.

Reply:
Sorry, I was not reading carefully. Nevermind. Resolving.